flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/nodes/dataset/forwarding/FieldForwardingUtils.scala

```diff
@@ -0,0 +1,143 @@
+import org.apache.flink.table.api.TableException
+import org.apache.flink.types.Row
+
+object FieldForwardingUtils {
+
+  def compositeTypeField = (fields: Seq[(String, TypeInfo[_])]) => fields
+
+  private def throwMissedWrapperException(customWrapper: TypeInfo[_]) = {
+    throw new TableException(s"Implementation for $customWrapper wrapper is missing.")
+  }
+
+  /**
+    * Wrapper for {@link getForwardedFields}
+    */
+  def getForwardedInput(
+      inputType: TypeInfo[_],
+      outputType: TypeInfo[_],
+      forwardIndices: Seq[Int],
+      customWrapper: TypeInfo[_] =>
+        Seq[(String, TypeInfo[_])] = throwMissedWrapperException): String = {
+
+    getForwardedFields(inputType,
+      outputType,
+      forwardIndices.zip(forwardIndices),
+      customWrapper)
+  }
+
+  /**
+    * Wraps provided indices with proper names.
+    * e.g. _1 for Tuple, f0 for Row, fieldName for POJO and named Row
+    *
+    * @param inputType      information of input data
+    * @param outputType     information of output data
+    * @param forwardIndices tuple of (input, output) indices of a forwarded field
+    * @param customWrapper  used for figuring out proper type in specific cases,
+    *                       e.g. {@see DataSetSingleRowJoin}
+    * @return string with forwarded fields mapped from input to output
+    */
+  def getForwardedFields(
+      inputType: TypeInfo[_],
+      outputType: TypeInfo[_],
+      forwardIndices: Seq[(Int, Int)],
+      customWrapper: TypeInfo[_] =>
+        Seq[(String, TypeInfo[_])] = throwMissedWrapperException): String = {
+
+    def chooseWrapper(
+        typeInformation: TypeInfo[_]): Seq[(String, TypeInfo[_])] = {
+
+      typeInformation match {
+        case composite: CompositeType[_] =>
+          val fields = extractFields(composite)
+          compositeTypeField(fields)
+        case basic: BasicTypeInfo[_] =>
+          Seq((s"*", basic))
+        case array: BasicArrayTypeInfo[_, _] =>
+          Seq((s"*", array))
+        case _ =>
+          customWrapper(typeInformation)
+      }
+    }
+
+    val wrapInput = chooseWrapper(inputType)
+    val wrapOutput = chooseWrapper(outputType)
+
+    forwardFields(forwardIndices, wrapInput, wrapOutput)
+  }
+
+  private def extractFields(
+      composite: CompositeType[_]): Seq[(String, TypeInfo[_])] = {
+
+    val types = for {
+      i <- 0 until composite.getArity
+    } yield { composite.getTypeAt(i) }
+
+    composite.getFieldNames.zip(types)
+  }
+
+  private def forwardFields(
+      forwardIndices: Seq[(Int, Int)],
+      wrappedInput: Int => (String, TypeInfo[_]),
+      wrappedOutput: Int => (String, TypeInfo[_])): String = {
+
+    implicit class Field2ForwardField(left: (String, TypeInfo[_])) {
+      def ->(right: (String, TypeInfo[_])): String = if (left.equals(right)) {
+        s"${left._1}"
+      } else {
+        if (left._2.equals(right._2)) {
+          s"${left._1}->${right._1}"
+        } else {
+          null
```

Comment:
I think this case should never happen if correct mappings are provided. If the types do not match, we have a bug the logic to create forward field mappings. If the mapping is not correct, but the types match by chance, we will have wrong forward fields which might cause the job to compute wrong results.

If the code enters the branch of non-matching types with the current tests, this indicates that there is a bug in the field mapping logic and this should not be ignored.